lib/router/src/cache.rs

```diff
@@ -1,426 +1,321 @@
+            values: IndexMap::default(),
+        };
+        let cache = Lock::new(cache);
+        let bg_purge = PurgeCache {
+            cache: cache.clone(),
+            interval: Interval::new_interval(expires),
+        };
 
-impl<'a, T: 'a, N: Now + 'a> Deref for Access<'a, T, N> {
-    type Target = T;
-    fn deref(&self) -> &Self::Target {
-        &self.node
+        (cache, bg_purge)
     }
-}
 
-impl<'a, T: 'a, N: Now + 'a> DerefMut for Access<'a, T, N> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.node
+    pub fn capacity(&self) -> usize {
+        self.capacity
     }
-}
 
-impl<'a, T: 'a, N: Now + 'a> Access<'a, T, N> {
-    #[cfg(test)]
-    fn last_access(&self) -> Instant {
-        self.node.last_access
+    pub fn can_insert(&self) -> bool {
+        self.values.len() < self.capacity
     }
-}
 
-impl<'a, T: 'a, N: Now + 'a> Drop for Access<'a, T, N> {
-    fn drop(&mut self) {
-        self.node.last_access = self.now.now();
-    }
-}
-
-// ===== impl Node =====
+    /// Attempts to access an item by key.
+    ///
+    /// If a value is returned, this key will not be considered for eviction
+    /// for another `expires` span of time.
+    pub fn access(&mut self, key: &K) -> Option<V> {
+        if let Some(node) = self.values.get_mut(key) {
+            self.expirations.reset(&node.dq_key, self.expires);
+            trace!("reset expiration for cache value associated with key");
+
+            return Some(node.value.clone());
+        }
 
-impl<T> Node<T> {
-    pub fn new(value: T, last_access: Instant) -> Self {
-        Node { value, last_access }
+        None
     }
 
-    pub fn access<'a, N: Now + 'a>(&'a mut self, now: &'a N) -> Access<'a, T, N> {
-        Access { now, node: self }
-    }
+    /// Attempts to insert an item by key.
+    ///
+    /// If a value is returned, this key has been set to expire after an
+    /// `expires` span of time.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        let node = {
+            trace!("inserting value into expirations");
+            let dq_key = self.expirations.insert(key.clone(), self.expires);
+            Node { dq_key, value }
+        };
 
-    pub fn last_access(&self) -> Instant {
-        self.last_access
+        trace!("inserting node into values");
+        self.values.insert(key, node).map(|n| n.value)
     }
-}
 
-impl<T> Deref for Node<T> {
-    type Target = T;
-    fn deref(&self) -> &Self::Target {
-        &self.value
-    }
-}
+    /// Evict expired values from the cache.
+    ///
+    /// Polls the underlying `DelayQueue`. When elements are returned from the
+    /// queue, remove the associated key from `values`.
+    fn poll_purge(&mut self) -> Poll<(), Error> {
+        while let Some(expired) = try_ready!(self.expirations.poll()) {
+            trace!("expired an entry from expirations");
+            self.values.remove(expired.get_ref());
+            trace!("removed an entry from values!");
```

Comment:
This log line is redundant given the immedately preceding log line.

Reply:
Its logging the fact that we are removing a value from both the internal `expirations`, and then router facing state of the cache `values`, but I see your point. I'll remove one